prover/src/constraints/evaluator/default.rs

```diff
@@ -131,18 +152,44 @@ where
         let transition_constraints =
             air.get_transition_constraints(&composition_coefficients.transition);
 
+        let lagrange_kernel_transition_constraints =
+            air.context().lagrange_kernel_aux_column_idx().map(|_| {
+                LagrangeKernelTransitionConstraints::new(
+                    air.context(),
+                    composition_coefficients.lagrange_kernel_transition,
+                )
+            });
+
         // build periodic value table
         let periodic_values = PeriodicValueTable::new(air);
 
         // build boundary constraint groups; these will be used to evaluate and compose boundary
         // constraint evaluations.
-        let boundary_constraints =
-            BoundaryConstraints::new(air, &aux_rand_elements, &composition_coefficients.boundary);
+        let boundary_constraints = BoundaryConstraints::new(
+            air,
+            &aux_rand_elements,
+            &composition_coefficients.boundary,
+            composition_coefficients.lagrange_kernel_boundary,
+        );
+
+        let lagrange_kernel_boundary_constraint = composition_coefficients
+            .lagrange_kernel_boundary
+            .map(|lagrange_kernel_boundary_coefficient| {
+                let lagrange_kernel_aux_rand_elements =
+                    air.lagrange_kernel_rand_elements(aux_rand_elements.get_segment_elements(0));
+
+                LagrangeKernelBoundaryConstraint::new(
+                    lagrange_kernel_boundary_coefficient,
+                    lagrange_kernel_aux_rand_elements,
+                )
+            });
```

Comment:
Instead of separating Lagrange boundary and transition constraints here, I wonder if we should treat them as a single unit - e.g., `LagrangeKernelConstraints` instantiated something like:
```Rust
let lagrange_kernel_constraints =
 air.context().lagrange_kernel_aux_column_idx().map(|_| {
     LagrangeKernelConstraints::new(
         air.context(),
         lagrange_rand_elements, // the r-values
         composition_coefficients.lagrange,
     )
 });
```
And then we split into evaluating Lagrange boundary and transition constraints internally within this struct.

I guess the main question is how to get the $r$-values here.

Reply:
I think we should instead generalize the existing `TransitionConstraints` and `BoundaryConstraints`, as detailed in https://github.com/facebook/winterfell/issues/261